Custom HTTP codes are not supported

**Describe the bug**
Spring Cloud Openfeign version: 4.1.0
Using `HttpStatus.valueof(int)` in [SpringDecoder](https://github.com/spring-cloud/spring-cloud-openfeign/blob/main/spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringDecoder.java#L86) and [ResponseEntityDecoder](https://github.com/spring-cloud/spring-cloud-openfeign/blob/main/spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/ResponseEntityDecoder.java#L87) causes an exception to be thrown if using HTTP codes that are not listed in HttpStatus.

```
Caused by: java.lang.IllegalArgumentException: No matching constant for [245]
	at org.springframework.http.HttpStatus.valueOf(HttpStatus.java:515) ~[spring-web-6.1.5.jar:6.1.5]
	at org.springframework.cloud.openfeign.support.SpringDecoder$FeignResponseAdapter.getStatusCode(SpringDecoder.java:86) ~[spring-cloud-openfeign-core-4.1.0.jar:4.1.0]
	at org.springframework.cloud.openfeign.support.SpringDecoder$FeignResponseAdapter.getStatusCode(SpringDecoder.java:76) ~[spring-cloud-openfeign-core-4.1.0.jar:4.1.0]
	at org.springframework.web.client.ClientHttpResponseDecorator.getStatusCode(ClientHttpResponseDecorator.java:55) ~[spring-web-6.1.5.jar:6.1.5]
	at org.springframework.web.client.IntrospectingClientHttpResponse.hasMessageBody(IntrospectingClientHttpResponse.java:60) ~[spring-web-6.1.5.jar:6.1.5]
	at org.springframework.web.client.HttpMessageConverterExtractor.extractData(HttpMessageConverterExtractor.java:90) ~[spring-web-6.1.5.jar:6.1.5]
	at org.springframework.cloud.openfeign.support.SpringDecoder.decode(SpringDecoder.java:70) ~[spring-cloud-openfeign-core-4.1.0.jar:4.1.0]
	at org.springframework.cloud.openfeign.support.ResponseEntityDecoder.decode(ResponseEntityDecoder.java:53) ~[spring-cloud-openfeign-core-4.1.0.jar:4.1.0]
	at feign.optionals.OptionalDecoder.decode(OptionalDecoder.java:36) ~[feign-core-13.1.jar:na]
```

**Sample**
producer application:
```
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@SpringBootApplication
public class ProducerApplication {
    public static void main(String[] args) {
        SpringApplication.run(ProducerApplication.class, args);
    }

    @GetMapping("/api/v1/entity")
    public ResponseEntity<String> getSomeEntity() {
        return ResponseEntity.status(245)
                .body("Producer - 245 status");
    }
}
```
consumer application:
```
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.cloud.openfeign.EnableFeignClients;
import org.springframework.http.ResponseEntity;

@SpringBootApplication
@EnableFeignClients
@RequiredArgsConstructor
public class ConsumerApplication implements CommandLineRunner {
    private static final Logger log = LoggerFactory.getLogger(ConsumerApplication.class);
    private final ProducerApiFeignClient producerClient;

    public static void main(String[] args) {
        SpringApplication.run(ConsumerApplication.class, args);
    }

    @Override
    public void run(String... args) {
        ResponseEntity<String> someEntity = producerClient.someEntity();
        log.info("Produce API - Http status code [{}]", someEntity.getStatusCode());
        log.info("Produce API - Body [{}]", someEntity.getBody());
    }
}
```
```
import org.springframework.cloud.openfeign.FeignClient;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;

@FeignClient(value = "producer-api-client", url = "http://localhost:8081")
public interface ProducerApiFeignClient {
    @GetMapping("/api/v1/entity")
    ResponseEntity<String> someEntity();
}
```

--- a/spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/ResponseEntityDecoder.java
+++ b/spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/ResponseEntityDecoder.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2013-2022 the original author or authors.
+ * Copyright 2013-2024 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -36,6 +36,7 @@ import org.springframework.http.ResponseEntity;
  *
  * @author chad jaros
  * @author Olga Maciaszek-Sharma
+ * @author Maksym Pasichenko
  */
 public class ResponseEntityDecoder implements Decoder {
 

--- a/spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/ResponseEntityDecoder.java
+++ b/spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/ResponseEntityDecoder.java
@@ -27,7 +27,7 @@ import feign.codec.Decoder;
 
 import org.springframework.http.HttpEntity;
 import org.springframework.http.HttpHeaders;
-import org.springframework.http.HttpStatus;
+import org.springframework.http.HttpStatusCode;
 import org.springframework.http.ResponseEntity;
 
 /**
@@ -84,7 +84,7 @@ public class ResponseEntityDecoder implements Decoder {
 			headers.put(key, new LinkedList<>(response.headers().get(key)));
 		}
 
-		return new ResponseEntity<>((T) instance, headers, HttpStatus.valueOf(response.status()));
+		return new ResponseEntity<>((T) instance, headers, HttpStatusCode.valueOf(response.status()));
 	}
 
 }

--- a/spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringDecoder.java
+++ b/spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringDecoder.java
@@ -42,6 +42,7 @@ import static org.springframework.cloud.openfeign.support.FeignUtils.getHttpHead
 /**
  * @author Spencer Gibb
  * @author Olga Maciaszek-Sharma
+ * @author Maksym Pasichenko
  */
 public class SpringDecoder implements Decoder {
 

--- a/spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringDecoder.java
+++ b/spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringDecoder.java
@@ -32,7 +32,7 @@ import org.springframework.beans.factory.ObjectFactory;
 import org.springframework.beans.factory.ObjectProvider;
 import org.springframework.boot.autoconfigure.http.HttpMessageConverters;
 import org.springframework.http.HttpHeaders;
-import org.springframework.http.HttpStatus;
+import org.springframework.http.HttpStatusCode;
 import org.springframework.http.client.ClientHttpResponse;
 import org.springframework.http.converter.HttpMessageConverter;
 import org.springframework.web.client.HttpMessageConverterExtractor;
@@ -82,8 +82,8 @@ public class SpringDecoder implements Decoder {
 		}
 
 		@Override
-		public HttpStatus getStatusCode() {
-			return HttpStatus.valueOf(response.status());
+		public HttpStatusCode getStatusCode() {
+			return HttpStatusCode.valueOf(response.status());
 		}
 
 		@Override

--- a/spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/SpringDecoderIntegrationTests.java
+++ b/spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/SpringDecoderIntegrationTests.java
@@ -156,7 +156,7 @@ class SpringDecoderIntegrationTests extends FeignClientFactoryBean {
 
 	@Test
 	// Issue: https://github.com/spring-cloud/spring-cloud-openfeign/issues/1010
-	void testResponseEntityWithCustomHttpStatusCode(){
+	void testResponseEntityWithCustomHttpStatusCode() {
 		ResponseEntity<Hello> response = testClient().getCustomHttpStatusCodeResponse();
 		assertThat(response.getStatusCode()).isEqualTo(HttpStatusCode.valueOf(245));
 	}
@@ -165,7 +165,7 @@ class SpringDecoderIntegrationTests extends FeignClientFactoryBean {
 
 		@GetMapping("/helloresponse")
 		ResponseEntity<Hello> getHelloResponse();
-		
+
 		@GetMapping("/hellocustomhttpstatuscode")
 		ResponseEntity<Hello> getCustomHttpStatusCodeResponse();
 

--- a/spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/SpringDecoderIntegrationTests.java
+++ b/spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/SpringDecoderIntegrationTests.java
@@ -35,6 +35,7 @@ import org.springframework.context.annotation.Configuration;
 import org.springframework.context.annotation.Import;
 import org.springframework.http.HttpHeaders;
 import org.springframework.http.HttpStatus;
+import org.springframework.http.HttpStatusCode;
 import org.springframework.http.MediaType;
 import org.springframework.http.ResponseEntity;
 import org.springframework.test.annotation.DirtiesContext;
@@ -49,6 +50,7 @@ import static org.assertj.core.api.Assertions.assertThat;
  * @author Spencer Gibb
  * @author Olga Maciaszek-Sharma
  * @author Szymon Linowski
+ * @author Maksym Pasichenko
  */
 @SpringBootTest(classes = SpringDecoderIntegrationTests.Application.class, webEnvironment = WebEnvironment.RANDOM_PORT,
 		value = { "spring.application.name=springdecodertest", "spring.jmx.enabled=false" })
@@ -152,10 +154,20 @@ class SpringDecoderIntegrationTests extends FeignClientFactoryBean {
 		assertThat(response.getHeaders().getContentType()).isEqualTo(MediaType.APPLICATION_JSON);
 	}
 
+	@Test
+	// Issue: https://github.com/spring-cloud/spring-cloud-openfeign/issues/1010
+	void testResponseEntityWithCustomHttpStatusCode(){
+		ResponseEntity<Hello> response = testClient().getCustomHttpStatusCodeResponse();
+		assertThat(response.getStatusCode()).isEqualTo(HttpStatusCode.valueOf(245));
+	}
+
 	protected interface TestClient {
 
 		@GetMapping("/helloresponse")
 		ResponseEntity<Hello> getHelloResponse();
+		
+		@GetMapping("/hellocustomhttpstatuscode")
+		ResponseEntity<Hello> getCustomHttpStatusCodeResponse();
 
 		@GetMapping("/hellovoid")
 		ResponseEntity<Void> getHelloVoid();
@@ -227,6 +239,11 @@ class SpringDecoderIntegrationTests extends FeignClientFactoryBean {
 		@Override
 		public ResponseEntity<Hello> getHelloResponse() {
 			return ResponseEntity.ok(new Hello("hello world via response"));
+		}
+
+		@Override
+		public ResponseEntity<Hello> getCustomHttpStatusCodeResponse() {
+			return ResponseEntity.status(245).body(new Hello("hello world via response with custom HTTP status code"));
 		}
 
 		@Override

